src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+section \<open>Chain Synchronization Mini-Protocol\<close>
+
+text \<open>
+  The theory in this section implements the chain synchronization mini-protocol.
+\<close>
+
+theory "Ouroboros-Mini_Protocols-Chain_Sync"
+  imports
+    "Ouroboros-Mini_Protocols"
+    "HOL-Library.BNF_Corec"
+begin
+
+locale chain_sync =
+  fixes initial_items\<^sub>c :: "'i list"
+  fixes initial_items\<^sub>s :: "'i list"
+  fixes candidate_points :: "'i list \<Rightarrow> 'p list"
+  fixes best_intersection_point :: "'i list \<Rightarrow> 'p list \<Rightarrow> 'p option"
+  fixes point_of :: "'i \<Rightarrow> 'p"
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Client |
+  Server
+
+subsection \<open>State Machine\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype ('i, 'p) message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>'i\<close> |
+  is_roll_backward: RollBackward \<open>'p\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>'p list\<close> |
+  is_intersect_found: IntersectFound \<open>'p\<close> |
+  is_intersect_not_found: IntersectNotFound
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Client" |
+  "agent_in_state' _ = Server"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (partial_case m of
+    FindIntersect _ \<Rightarrow> Intersect |
+    RequestNext \<Rightarrow> CanAwait
+  )" |
+  "next_state' Intersect m = (partial_case m of
+    IntersectFound _ \<Rightarrow> Idle |
+    IntersectNotFound \<Rightarrow> Idle
+  )" |
+  "next_state' CanAwait m = (partial_case m of
+    RollForward _ \<Rightarrow> Idle |
+    RollBackward _ \<Rightarrow> Idle |
+    AwaitReply \<Rightarrow> Idle \<comment> \<open>only for this initial implementation\<close>
+  )"
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_items_update: ItemsUpdate
+
+definition roll_back_to :: "'i list \<Rightarrow> 'p \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> 'i list" where
+  [simp]: "roll_back_to \<I> p \<psi> = (THE \<I>\<^sub>1. \<exists>\<I>\<^sub>2. \<I> = \<I>\<^sub>1 @ \<I>\<^sub>2 \<and> \<psi> (last \<I>\<^sub>1) = p)"
+
+corec client_program where
+  "client_program ph \<kappa> \<psi> \<I> = (
+    case ph of
+      IntersectionFinding \<Rightarrow>
+        \<up> Cont (FindIntersect (\<kappa> \<I>));
+        \<down> M; (partial_case M of
+          Cont IntersectNotFound \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
```

Comment:
That's a good question. _The Shelley Networking Protocol_ does not say what the client should do in such a situation, although I would also agree with you in that it should bootstrap from genesis. However, if my understanding is correct, in the implementation the client terminates when no intersection point is found. I think we should consult @dcoutts about this.

Reply:
I just sent the following message to @dcoutts outside of GitHub:

> How does a chain-sync client react to an `IntersectNotFound` message? Javier says that, according to his understanding, the Haskell implementation let’s the client just terminate the protocol in this case. However, we both think that this is wrong. Shouldn’t the client just bootstrap from genesis in this case?

Reply:
I got an answer from @dcoutts.

Apparently, the Haskell implementation of the client indeed bails out if an intersection couldn’t be found. This behavior is acceptable if the client already has a chain and its chain is not so new that it would have only *K* blocks or less. In this case, absence of an intersection would mean that the server node is on the wrong chain according to the client and thus the client shouldn’t further consider it.

I think that for now we should stick to bailing out upon `IntersectNotFound` also in our specification. In the longer run, we should anyhow generalize our specification to cover more behavior that might not occur with the current implementation of the protocol but is nevertheless allowed by the protocol. Then we should also model different reactions to `IntersectNotFound`.

Reply:
Thanks for letting me know. Indeed, in the real implementation a point can either refer to the genesis block or to an ordinary block.

Reply:
Meanwhile, I’m not even sure whether we should assume that the client and the server possess the same genesis block. I had assumed that so far because of the following statement from *The Shelley Networking Protocol*:

> The chain producer assumes that a consumer, which has just connected, only knows the genesis block and initialises the read-pointer of that consumer with a pointer to the genesis block on its chain.

That said, even if the implementation makes sure outside of the chain synchronization protocol that the two parties use the same genesis block, we can still generalize our specification to also cope with the situation that they don’t.

I suggest that we change the specification such that the client chain and the server chain contain the respective genesis items, which may be equal or not. With this in place, points to genesis items are a natural thing, as they are just points to elements of chains like all other points. If we make this change, we shall add assumptions to the `chain_sync` locale that state that the initial client chain and the initial server chain are non-empty.

Reply:
Fixed in f9ee76c6dd21108be7a92ab49d9ed10a03f9ccb0 and 52d2ec2f9818b287f58ee712a7159f1e93cdfd93, please check.

Reply:
Regarding f9ee76c6dd21108be7a92ab49d9ed10a03f9ccb0: The assumptions should get names. I usually write locale assumptions such that I put `assumes ⟨name⟩:` on a line and put the actual assumption on the following line, indenting it relative to the `assumes` line.

Reply:
Fixed in 8df2f814c6a34b6ebc820d37df0a3df5da062002, please check.

Reply:
Regarding 52d2ec2f9818b287f58ee712a7159f1e93cdfd93:

I think that, in the corecursive call of the `if b` branch, you have to change the argument `Suc k` to `k`.

Reply:
You're right, good catch! Fixed in 05db295caa6b934441a24277a411df8351ddedce, please check.

Reply:
> Fixed in [8df2f81](https://github.com/input-output-hk/ouroboros-high-assurance/commit/8df2f814c6a34b6ebc820d37df0a3df5da062002), please check.

I wonder whether `genesis_item_in_initial_[…]_chain` is a good choice of name. Yes, with the background of our discussion, it makes perfect sense, since the non-emptiness requirement arises from us now putting genesis items into these lists as well. However, the non-emptiness requirement as such doesn’t say in any way that one of the items in such a list must be a genesis item. Therefore, I suggest to just use names of the form `initial_[…]_chain_is_not_empty`. What do you think?

Reply:
Agreed and fixed in 7dd7f67484f8f3cc7c6c893ba93d939f233d7aba, please check.